### ISSUE TITLE
About VirtualFile.py?

### ISSUE THREAD

**Issue by HymEric** [LoSealL/VideoSuperResolution#37]
In this repo, VSR.Dataloader.VirtualFile is complicate, can you tell me why there is VirtualFile.py? I found this project is complicate, can you tell me some ideas or guidelines to study it because I want to learn deeply.

**Comment by LoSealL** [comment]
-------------------------------------------------------
That's ... also difficult for me :(...
I am sticking to write documents and tutorials for this project, but couldn't find my time... I apologize for this, and will try my best to explain here briefly.

## DataLoader
Data loader is one of the core contributions of VSR. We have to face much more complicated data structures in multiple low-level CV tasks other than SISR, and data could be too large to read all of them into memory. So data loader here is a fast and memory efficient I/O system. It doesn't need to build binary files such as `HDF5` or `lmdb`, but can directly read data from structured folders. The data loader will build an index for all the files and generate patches for each epoch. Which makes us easy to add/remove/change or combine subsets of data.

`VirtualFile` is just **virtual file**, for it can read a single image, a sequence of images, or raw video data . It's the basis of data loader, making manipulating of different types of data simultaneously.

## Trainer
Trainer is the way to optimize model. There are some pre-defined trainers (such as VSRTrainer, GANTrainer...). A trainer is responsible for fetching data from data loader, and feed them into the model.

The main DIY part of trainer is the `train`, `benchmark` and `infer` implementation.

## Models
Models is the way to build graph, how to calculate loss, given input images and labels.

## Training and testing
For `tensorflow` (`VSR`), the training script is in `Train/run.py`. For `pytorch` (`VSRTorch`) they are `VSRTorch/train.py` and `VSRTorch/eval.py`. The options and arguments are a little different.

## Tools
`Tools` contains some independent useful scripts for research. Such as video transcoding (FFmpegHelper.py), image-video transforming, huge image dividing and combining (NtireHelper.py) and datasets pre-processing (DND.py, MakeHDF.py and CelebA.py)

## Others
For other parts, they depends on the DL framework.
### Tensorflow
TF is the original and the first choice of mine. For tensorflow, I try to simplify the codes needed to build model. There're lots of alias and helper functions. For example, you can use
```python
self.bn_relu_conv2d(x)
```
to construct a convolutional layer with batch normalization and ReLU activation.


### PyTorch
Torch is the 2nd and currently the major choice. `VSRTorch` shares the data loader to `VSR`, and the training structure is almost the same as `VSR`, except for how to build models.

**Comment by HymEric** [comment]
Thanks for your explanation and your great work. I'm looking forward to your documents and tutorials.
